File: 零碳园区优化_v6/carbon_accounting.m
function out = carbon_accounting(data, Pgrid, Pchp, Hchp, Fgas)
%CARBON_ACCOUNTING Compute carbon emission and free allowance.
% Units:
%   Pgrid, Pchp, Hchp, Fgas: MW or MWh/h over one time step
%   emission/quota outputs: kgCO2 per time step after multiplying by dt

dt = data.dt;
if nargin < 3 || isempty(Pchp)
    Pchp = zeros(size(Pgrid));
end
if nargin < 4 || isempty(Hchp)
    Hchp = zeros(size(Pgrid));
end
if nargin < 5 || isempty(Fgas) %#ok<INUSD>
    Fgas = zeros(size(Pgrid)); %#ok<NASGU>
end

zetaE = get_field(data, 'zetaE', 1080);    % kg/MWh, paper: 1.08 kg/kWh
zetaH = get_field(data, 'zetaH', 324);     % kg/MWh_th, paper: 0.09 t/GJ
chiE = get_field(data, 'chiE', 728);       % kg/MWh, paper: 0.728 kg/kWh
chiH = get_field(data, 'chiH', 367.2);     % kg/MWh_th, paper: 0.102 t/GJ
ceh = get_field(data, 'ceh', 1.6667);

chpHeatEquivalent = ceh .* Pchp + Hchp;

out.gridEmission = zetaE .* Pgrid .* dt;
out.chpEmission = zetaH .* chpHeatEquivalent .* dt;
out.emission = out.gridEmission + out.chpEmission;

out.gridQuota = chiE .* Pgrid .* dt;
out.chpQuota = chiH .* chpHeatEquivalent .* dt;
if isfield(data, 'fixedCarbonQuota_kg') && ~isempty(data.fixedCarbonQuota_kg)
    out.quota = data.fixedCarbonQuota_kg;
    out.gridQuota = zeros(size(out.quota));
    out.chpQuota = out.quota;
else
    out.quota = out.gridQuota + out.chpQuota;
end
out.netAllowanceSurplus = out.quota - out.emission;
end

function val = get_field(s, name, defaultVal)
if isfield(s, name) && ~isempty(s.(name))
    val = s.(name);
else
    val = defaultVal;
end
end

File: realtime_optimizer/carbon_accounting.py
import numpy as np


def compute_carbon(Pgrid, Pchp, Hchp, Fgas, data):
    dt = data['dt']
    zetaE = data.get('zetaE', 1080)
    zetaH = data.get('zetaH', 324)
    chiE = data.get('chiE', 728)
    chiH = data.get('chiH', 367.2)
    ceh = data.get('ceh', 1.6667)

    chp_heat_eq = ceh * Pchp + Hchp

    grid_emission = zetaE * Pgrid * dt
    chp_emission = zetaH * chp_heat_eq * dt
    emission = grid_emission + chp_emission

    grid_quota = chiE * Pgrid * dt
    chp_quota = chiH * chp_heat_eq * dt

    carbon_quota_mode = data.get('carbonQuotaMode', 'baseline-scenario')
    fixed_quota = data.get('fixedCarbonQuota_kg', None)
    use_fixed = (carbon_quota_mode == 'fixed-baseline') and (fixed_quota is not None)

    if use_fixed:
        quota = fixed_quota
    else:
        quota = grid_quota + chp_quota

    net_surplus = quota - emission

    return {
        'gridEmission': grid_emission,
        'chpEmission': chp_emission,
        'emission': emission,
        'gridQuota': grid_quota,
        'chpQuota': chp_quota,
        'quota': quota,
        'netAllowanceSurplus': net_surplus,
    }
